ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala

```diff
@@ -122,9 +124,25 @@ object SqlLedgerReaderWriter {
     for {
       uninitializedDatabase <- Database.owner(jdbcUrl)
       database = uninitializedDatabase.migrate()
+      queries = database.queries
+      ledgerId <- ResourceOwner.forFuture(() =>
+        database.inWriteTransaction("Checking ledger ID at startup") { implicit connection =>
+          val providedLedgerId =
+            initialLedgerId.getOrElse(Ref.LedgerString.assertFromString(UUID.randomUUID.toString))
+          val ledgerId = queries.updateOrRetrieveLedgerId(providedLedgerId)
+          if (initialLedgerId.exists(_ != ledgerId)) {
+            Future.failed(
+              new LedgerIdMismatchException(
+                domain.LedgerId(ledgerId),
+                domain.LedgerId(initialLedgerId.get),
+              ))
+          } else {
+            Future.successful(ledgerId)
+          }
+      })
       head <- ResourceOwner.forFuture(() =>
         database.inReadTransaction("Reading head at startup") { implicit connection =>
-          Future(database.queries.selectLatestLogEntryId().map(_ + 1).getOrElse(StartIndex))
+          Future.successful(queries.selectLatestLogEntryId().map(_ + 1).getOrElse(StartIndex))
```

Comment:
This operation and the previous one can probably be composed in a `Future` as part of another method so that we don't have to wrap in a `ResourceOwner` twice over two steps.

Reply:
True, but then we have to return a tuple, which is less pleasant. I'm happy to combine this one with the next one, for the dispatcher.